src/transform/src/analysis/equivalences.rs

```diff
@@ -0,0 +1,582 @@
+pub struct EquivalenceClasses {
+    /// Multiple lists of equivalent expressions, each representing an equivalence class.
+    ///
+    /// The first element should be the "canonical" simplest element, that any other element
+    /// can be replaced by.
+    /// These classes are unified whenever possible, to minimize the number of classes.
+    pub classes: Vec<Vec<MirScalarExpr>>,
+}
+
+impl EquivalenceClasses {
+    /// Sorts and deduplicates each class, and the classes themselves.
+    fn tidy(&mut self) {
+        for class in self.classes.iter_mut() {
+            // Remove all literal errors, as they cannot be equated to other things.
+            class.retain(|e| !e.is_literal_err());
+            class.sort_by(|e1, e2| match (e1, e2) {
+                (MirScalarExpr::Literal(_, _), MirScalarExpr::Literal(_, _)) => e1.cmp(e2),
+                (MirScalarExpr::Literal(_, _), _) => std::cmp::Ordering::Less,
+                (_, MirScalarExpr::Literal(_, _)) => std::cmp::Ordering::Greater,
+                (MirScalarExpr::Column(_), MirScalarExpr::Column(_)) => e1.cmp(e2),
+                (MirScalarExpr::Column(_), _) => std::cmp::Ordering::Less,
+                (_, MirScalarExpr::Column(_)) => std::cmp::Ordering::Greater,
+                (x, y) => {
+                    // General expressions should be ordered by their size,
+                    // to ensure we only simplify expressions by substitution.
+                    let x_size = x.size();
+                    let y_size = y.size();
+                    if x_size == y_size {
+                        x.cmp(y)
+                    } else {
+                        x_size.cmp(&y_size)
+                    }
+                }
+            });
+            class.dedup();
+        }
+        self.classes.retain(|c| c.len() > 1);
+        self.classes.sort();
+        self.classes.dedup();
+    }
+
+    /// Update `self` to maintain the same equivalences which potentially reducing along `Ord::le`.
+    ///
+    /// Informally this means simplifying constraints, removing redundant constraints, and unifying equivalence classes.
+    pub fn minimize(&mut self, columns: &Option<Vec<ColumnType>>) {
+        // Repeatedly, we reduce each of the classes themselves, then unify the classes.
+        // This should strictly reduce complexity, and reach a fixed point.
+        // Ideally it is *confluent*, arriving at the same fixed point no matter the order of operations.
+        self.tidy();
+
+        // We continue as long as any simplification has occurred.
+        // An expression can be simplified, a duplication found, or two classes unified.
+        let mut stable = false;
+        while !stable {
+            stable = self.minimize_once(columns);
+        }
+
+        // TODO: remove these measures once we are more confident about idempotence.
+        let prev = self.clone();
+        self.minimize_once(columns);
+        assert_eq!(self, &prev);
```

Comment:
Copying here from Slack:
I'd suggest to leave the idempotence assertion in the code (as a `soft_assert_or_log`) until we test this on customer queries.